Fixes github rate limit

rin-st opened the pull request:
Everything is already described inside #260, I just wrote the code. Thanks for suggestions!

Instead of calling github API I used two calls to github.com using Promise.all

Test examples:
with both solidity frameworks, options will be hardhat and foundry
```
yarn cli --skip -e erc-20
```

with one solidity framework (hh), no options and autoinstall
```
yarn cli --skip -e challenge-0-simple-nft
```

with no solidity framework, options are hardhat, foundry and none
```
yarn cli --skip -e subgraph
```

Fixes #260

technophile-04 reviewed (approved):
Tysm @rin-st!! I think this is great! 

Just extending your main comment: 

If the github request fails, we show the all three options to user (hardhat, foundry, none) which is good! 


Just cc @carletex shall we log some warning or we just let it fail silently? I think failing silently is fine? I mean now with github.com its very unlikely to fail so it's fine if we don't log the warning or anything?

carletex commented:
> Just cc @carletex shall we log some warning or we just let it fail silently? I think failing silently is fine? I mean now with github.com its very unlikely to fail so it's fine if we don't log the warning or anything?

I think a warning won't hurt, but no big deal for me, so up to y'all

rin-st commented:
Added warning

From initial issue message:
> Just extending it ahead, if we get rate limited in CLI interface we show user all three options (hardhat, foundry, none) it's fine if extension doesn't provide foundry we have default foundry and extension should at least work (this is how CLI used work earlier). Maybe we show a small warning or something that this happened and check on gh to select solidity framework which is only given by extension or something.

So also added possibility to choose the framework if framework check fails

technophile-04 commented on src/utils/external-extensions.ts:
```diff
@@ -165,13 +165,27 @@ export const getSolidityFrameworkDirsFromExternalExtension = async (
+  const frameworkChecks = solidityFrameworks.map(async framework => {
+    const githubUrl = `https://github.com/${ownerName}/${repoName}/tree/${branch}/extension/packages/${framework}`;
+    try {
+      const res = await fetch(githubUrl);
+      if (res.ok) {
+        return framework as SolidityFramework;
+      }
+      // Directory doesn't exist
+      return null;
```
I think we should be deterministic here. 

```ts
If (res.status === 200) return framework as SolidityFramework;

// Actually directory not found
if(res.status === 404) return null;

// Any status code other than that we should show console warning. This can rate limit code (429 or 403) or any other code
throw new error (`${framework.charAt(0).toUpperCase() + framework.slice(1)} framework check failed with status `${res.status}`. You can verify it at ${githubUrl}.`)
```

And in catch we could just show message console.warn(error.message) and then return framework 

what do you think? 

Saying this because currently if we get rate limiting we want log the warning. 

--- 

Here are github API status codes: 

![Screenshot 2025-05-23 at 3 05 59 PM](https://github.com/user-attachments/assets/a2e72289-45e6-40e7-83cd-8328894bb992)
![Screenshot 2025-05-23 at 3 06 15 PM](https://github.com/user-attachments/assets/db2c00dd-f148-444e-ac65-083fa9ae0347)

this are for api.github.com but shall be similar not found and rate limiting status code will be same mostly.

rin-st replied:
Agree, thank you! [ea79140](https://github.com/scaffold-eth/create-eth/pull/264/commits/ea79140dfaa79756c6c1cdd96385e5426007cab7)

technophile-04 commented:
Tysm @rin-st works great!1
